File: largest_rectangle_in_histogram.cpp
#include "header.h"

//http://fisherlei.blogspot.com/2012/12/leetcode-largest-rectangle-in-histogram.html
//http://www.geeksforgeeks.org/largest-rectangle-under-histogram/

void find_left(vector<int>& height, vector<int>& v){
    if(height.size() == 0) return;
    stack<int> s;
    height.push_back(0);
    for(int i=0; i<height.size(); i++){
        if(s.empty() || height[s.top()]<height[i]){
            s.push(i);
            continue;
        }
        while(!s.empty() && height[s.top()]>=height[i]){
            int x = s.top();
            s.pop();
            v[x] = s.empty()?height[x]*(x+1):height[x]*(x-s.top());
        }
        s.push(i);
    }
    height.pop_back();
}

int largestRectangleArea(vector<int>& height){
    vector<int> left(height.size(), 0);
    vector<int> right(height.size(), 0);
    find_left(height, left);
    reverse(height.begin(), height.end());
    find_left(height, right);
    reverse(right.begin(), right.end());
    reverse(height.begin(), height.end());
    int res = 0;
    for(int i=0; i<left.size(); i++)
        res = max(res, left[i]+right[i]-height[i]);
    return res;
}

int main(){
//    int height[] = {6, 2, 5, 4, 5, 1, 6};
    int height[] = {2,4};
    vector<int> h_v(height,height+sizeof(height)/sizeof(height[0]));
    printf("%d\n", largestRectangleArea(h_v));
    return 0;
}

File: largest_rectangle_in_histogram.py
class Solution:
    def get_left(self,height):
        left = [0 for i in range(len(height))]
        height.append(0)
        stack = []
        for i in range(0,len(height),1):
            if stack == []:
                stack.append(i)
                continue
            while stack != [] and height[stack[-1]]>=height[i]:
                n = stack.pop()
                left[n] = height[n]*(n+1 if stack == [] else n-stack[-1])
            stack.append(i)
        height.pop()
        return left
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height):
        left = self.get_left(height)
        right = list(reversed(self.get_left(list(reversed(height)))))
        res = 0 
        for i in range(len(height)):
            res = max(res,left[i]+right[i]-height[i])
        return res
        
print Solution().largestRectangleArea([2,1,5,6,2,3])
